GitUI/CommandsDialogs/RevisionDiffControl.cs

```diff
@@ -948,5 +973,372 @@ public void SwitchFocus(bool alreadyContainedFocus)
+        }
+
+        private void DiffText_TextLoaded(object sender, EventArgs e)
+        {
+            _selectedDiffReloaded = true;
+
+            var item = DiffFiles?.SelectedItem;
+            if (item == null)
+            {
+                return;
+            }
+
+            // Show the menu texts etc for the currently displayed file
+            if (IsWorkTreeOrIndexWithParent())
+            {
+                // HEAD -> Index, Index -> Worktree allows unstaging/staging
+                if (item.SecondRevision.ObjectId == ObjectId.IndexId)
+                {
+                    DiffText.CherryPickContextMenuEntry_Update(
+                        Strings.UnstageSelectedLines,
+                        Properties.Images.Unstage,
+                        GetShortcutKeyDisplayString(Command.UnStageSelectedFile));
+                }
+                else
+                {
+                    DiffText.CherryPickContextMenuEntry_Update(
+                        Strings.StageSelectedLines,
+                        Properties.Images.Stage,
+                        GetShortcutKeyDisplayString(Command.StageSelectedFile));
+                }
+
+                DiffText.RevertSelectedContextMenuEntry_Update(
+                    Strings.ResetSelectedLines,
+                    Properties.Images.ResetWorkingDirChanges,
+                    GetShortcutKeyDisplayString(Command.ResetSelectedFiles));
+
+                if (item.Item.IsNew && !FileHelper.IsImage(item.Item.Name))
+                {
+                    // Add visibility, not for normal file displays
+                    DiffText.CherryPickContextMenuEntry_Visible();
+                    DiffText.RevertSelectedContextMenuEntry_Visible();
+                }
+
+                return;
+            }
+
+            DiffText.CherryPickContextMenuEntry_Update(
+                Strings.CherrypickSelectedLines,
+                Properties.Images.CherryPick,
+                GetShortcutKeyDisplayString(Command.StageSelectedFile));
+
+            DiffText.RevertSelectedContextMenuEntry_Update(
+                Strings.RevertSelectedLines,
+                Properties.Images.ResetFileTo,
+                GetShortcutKeyDisplayString(Command.ResetSelectedFiles));
+        }
+
+        /// <summary>
+        /// Check if the commits for the currenly selected item is the same as in FormCommit,
+        /// i.e. HEAD->Index or Index->WorkTree
+        /// </summary>
+        /// <returns>If 'artificial' handling</returns>
+        private bool IsWorkTreeOrIndexWithParent()
+        {
+            var item = DiffFiles.SelectedItem;
+            if (item == null || !item.SecondRevision.IsArtificial)
+            {
+                return false;
+            }
+
+            return (item.SecondRevision.ObjectId == ObjectId.WorkTreeId
+                    || item.SecondRevision.ObjectId == ObjectId.IndexId)
+                   && item.SecondRevision.FirstParentGuid == item.FirstRevision?.ObjectId;
```

Comment:
Do you still consider this WIP -- except this minor merge conflict?
```suggestion
                   && item.SecondRevision.FirstParentId == item.FirstRevision?.ObjectId;
```

Reply:
This is high priority for me for next release. I get very frustrated when I am not using this and have to use FormCommit...

RussKie requested changes, he was not satisfied with keeping the responsibilities as they are before this PR. 
I therefore have have to rewrite some, which I have not started with yet. (Wanted the changes leading up to 3.4 to settle and not spend all time rebasing, then I want to finish other tasks that are closer to completion.)

Reply:
I had already changed this in my wip branch...